### PR TITLE
fix: Close dropdown on click

### LINKED ISSUE #3452: "More..." dropdown list is not closed if "Dictionaries & Language Packs" is clicked
Steps to reproduce:
1. Load AMO-dev page on your device https://addons-dev.allizom.org/en-US/android/
2. Navigate to "Dictionaries & Language Packs"

Expected results:
Corresponding page is loaded. The drop down list is closed.

Actual results:
Corresponding page is loaded. The drop down list is not closed.

Notes/Issues:
- reproducing on desktop too, while the hover is still on the option;
Verified on AMO-dev FF56(Win 7).
Video for this issue:
videotogif_2017 10 12_10 13 01

### PR DESCRIPTION
**Pull request by tofumatt**
Fix #3452.

Changes to use a JS hover to allow us to dismiss the menu on a click.

### CODE REVIEW

File: src/amo/components/SectionLinks/styles.scss [hunk 1]
```diff
@@ -67,7 +67,7 @@
   .DropdownMenu-button {
     display: block;
     margin-top: 2px;
-    padding: 0;
+    padding: 0 0 2px;
```
Comment (tofumatt) [inline src/amo/components/SectionLinks/styles.scss]: This prevents any issues when moving the mouse from the button to the menu–if this isn't here it might cause a `mouseLeave` event to be fired and the menu will disappear.

**Comment by muffinresearch** [comment]
Was there a reason to not use onClickOutside?

**Comment by willdurand** [comment]
> Was there a reason to not use onClickOutside?

isn't it because we are actually inside the component? This HOC is used if I remember correctly.

**Comment by tofumatt** [comment]
Yeah, we're clicking inside the component. The issue is clicks *inside* the component on a link that should dismiss the dropdown (and *would* in tradition navigation because it would reload the page, resetting the dropdown state) don't dismiss it so it hangs out there until the user hovers away. It's even worse on mobile where the dropdown takes up a lot of screen and there is no hover, so a user needs to click "Language Tools" and then click elsewhere to get it to dismiss.

File: src/ui/components/DropdownMenu/index.js [hunk 2]
```diff
@@ -34,18 +36,44 @@ export class DropdownMenuBase extends React.Component<Props, State> {
     }));
   }
 
+  handleOnClickForLinks = (event: SyntheticEvent<any>) => {
+    // If a link inside the menu is clicked, we should close the dropdown.
+    // See: https://github.com/mozilla/addons-frontend/issues/3452
+    if (event.target && event.target.tagName === 'A') {
```
Comment (willdurand) [inline src/ui/components/DropdownMenu/index.js]: Are we 100% sure it is a uppercase tag name? I think so but we never know.

File: src/ui/components/DropdownMenu/styles.scss [hunk 3]
```diff
@@ -11,22 +11,6 @@
...
-      .DropdownMenu-items {
-        display: block;
-      }
-
-      .Icon-inverted-caret {
-        opacity: 1;
-      }
-    }
-  }
-
```
Comment (willdurand) [inline src/ui/components/DropdownMenu/styles.scss]: Why is this removed? Not needed anymore?

**Comment by willdurand** [comment]
STRs:

- go with the mouse on the menu button/title
- click on it

This patch in action:



It actually closes the dropdown menu. It is opened on hover, but when you click on the menu button/title, it is closed.

- Current behavior on -dev: (I click many times when the cursor is on the button/title)



Same steps keep the dropdown menu open.

File: src/ui/components/DropdownMenu/styles.scss (hunk 3 above)
Comment (tofumatt) [reply src/ui/components/DropdownMenu/styles.scss]: Yeah, because the hover state is now managed in JS.

**Comment by tofumatt** [comment]
Actually, that was intentional (there is even a test case for it: https://github.com/mozilla/addons-frontend/pull/3632/files#diff-d411d62dcb1f4ae57c93cac125dba2c8R99).

This means a user can hide/show the menu with a click, which they may want to do. I thought there was no harm in it as the main menu items are never clickable links but just show/hide actions.

**Comment by willdurand** [comment]
hmm does that still work on mobile devices?

I actually triggered this bug the first time I interacted with the menu, because I am never sure of the `hover` behavior. I tend to click to things to keep them open.

**Comment by tofumatt** [comment]
It still works on mobile because the first click opens it. I guess I could make it so that hover sets a different state and then the first click doesn't do anything. But another click would still close it? Would that work for you?

**Comment by willdurand** [comment]
If that's quick and easy, yes. Otherwise let's forget about it.

**Comment by muffinresearch** [comment]
> Yeah, we're clicking inside the component. The issue is clicks inside the component on a link that should dismiss the dropdown (and would in tradition navigation because it would reload the page, resetting the dropdown state) don't dismiss it so it hangs out there until the user hovers away. It's even worse on mobile where the dropdown takes up a lot of screen and there is no hover, so a user needs to click "Language Tools" and then click elsewhere to get it to dismiss.

Aha right, that makes perfect sense.

**Comment by willdurand** [comment]
@tofumatt anything I can do here?

File: src/ui/components/DropdownMenu/index.js (hunk 2 above)
Comment (tofumatt) [reply src/ui/components/DropdownMenu/index.js]: That's how the DOM API works and React's `SyntheticEvent` normalizes the content across browsers, so it should always be uppercase.

**Comment by tofumatt** [comment]
I don't there's anything needing to be done other than a review? Does it look good to you or is there still stuff that needs addressing?

**Comment by willdurand** [comment]
I was waiting for a reply to:

> [tofumatt] I guess I could make it so that hover sets a different state and then the first click doesn't do anything. But another click would still close it? Would that work for you?

> [willdurand] If that's quick and easy, yes. Otherwise let's forget about it.

**Comment by tofumatt** [comment]
Oh, shoot, right. Okay I can do that.

**Comment by tofumatt** [comment]
@willdurand Okay, I've patched it so the first click after a hover won't close the menu... but clicks afterward do, which I think is sensible behaviour.
